posts/change-data-capture-cloudsql-postgres-with-apache-airflow-1.md

```diff
@@ -0,0 +1,1002 @@
+
+        @apply_defaults
+        def __init__(self, gcs_bucket, gcs_keypath, database_instance,
+                    database_name, export_sql, project_id,
+                    gcp_conn_id='gcs_default', offload=False, 
+                    *args, **kwargs):
+            
+            super().__init__(*args, **kwargs)
+            
+            self.gcp_conn_id = gcp_conn_id
+            self.gcs_bucket = gcs_bucket
+            self.gcs_keypath = gcs_keypath
+            self.database_instance = database_instance
+            self.database_name = database_name
+            self.export_sql = export_sql
+            self.project_id = project_id
+            self.offload = offload
+            self.file_type = 'CSV'
+
+        def execute(self, context):
+            try:
+                assert self.file_type == 'CSV'
+            except AssertionError:
+                raise AirflowException(f'Extract file_type must be CSV')
+
+            hook = CloudSqlHook(api_version='v1beta4', gcp_conn_id=self.gcp_conn_id)
+
+            newlines = re.compile("[\r\n]+(\s+|)")
+            
+    				# The query must have newlines removed
+            self.export_sql = re.sub(newlines, ' ', self.export_sql)\
+                                .strip()
+
+            body = {
+                "exportContext":
+                {
+                  "fileType": f"{self.file_type}",
+                  "uri": f"gs://{self.gcs_bucket}/{self.gcs_keypath}",
+                  "databases": [f"{self.database_name}"], 
+                  "offload": self.offload, # True for serverless (takes longer but offloads - use true for historical extract)
+                  "csvExportOptions":
+                   {
+                       "selectQuery": f"{self.export_sql}"
+                   }
+                }
+            }
+            hook.export_instance(instance=self.database_instance,
+                                 body=body, 
+                                 project_id=self.project_id)
+    ```
+
+2. Save your new custom operator. Note how some the parameters in the API call indicate this is configurable. You can now use the same code with a different values for `gcs_bucket`, `gcs_keypath`, `export_query`, `project_id`, and `gcp_connection`, making this operator reusable for any variation of this workflow. You can even modify it to fulfill a different component of the same workflow. 
+
+## Step 9: Implement Watermarking and Execution Dates
+
+In data synchronization, a [watermark can be defined as](https://en.wikipedia.org/wiki/Watermark_(data_synchronization)):
+
+an object of a predefined format which provides a point of reference for two systems/datasets attempting to establish delta/incremental synchronization; any object in the queried data source which was created, modified/changed, and/or deleted after the watermark value was established will be qualified as "above watermark" and could/should be returned to a delta-querying partner
```

Comment:
Capitalize this first letter and throw a period at the end here. Also, we should throw quotes around this (and maybe blockquote it?)

Reply:
good catch. that was a notion formatting drop